test/model.discriminator.test.js

```diff
@@ -1686,4 +1686,35 @@ describe('model', function() {
 
     assert.deepEqual(nestedDocument.body.children[1].body.children[0].body.children[0].body.children, []);
   });
+
+  describe('Discriminator Key test', function() {
+    it('(gh-9015)', function() {
+      return co(function*() {
+        const baseSchema = new Schema({}, { discriminatorKey: 'type' });
+        const baseModel = db.model('thing', baseSchema);
+        const aSchema = new Schema(
+          {
+            aThing: { type: Number }
+          },
+          { _id: false, id: false }
+        );
+        baseModel.discriminator('A', aSchema);
+        const bSchema = new Schema(
+          {
+            bThing: { type: String }
+          },
+          { _id: false, id: false }
+        );
+        baseModel.discriminator('B', bSchema);
+        // Model is created as a type A
+        const doc = yield baseModel.create({ type: 'A', aThing: 1 });
+        const res = yield baseModel.findByIdAndUpdate(
+          doc._id,
+          { type: 'B', bThing: 'one', aThing: '2' },
+          { runValidators: true, /* overwriteDiscriminatorKey: true, */ new: true }
+        );
```

Comment:
Add another `findByIdAndUpdate()` call to this test that shows using `overwriteDiscriminatorKey: true` changes `res.type`